crates/network-scheduler/src/assignment.rs

```diff
@@ -0,0 +1,235 @@
+use flate2::read::GzDecoder;
+use std::{collections::HashMap, io::Read};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+use crate::signature::timed_hmac_now;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Chunk {
+    pub id: String,
+    pub base_url: String,
+    pub files: HashMap<String, String>,
+    size: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Dataset {
+    pub id: String,
+    pub base_url: String,
+    pub chunks: Vec<Chunk>,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct EncryptedHeaders {
+    worker_id: String,
+    worker_signature: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WorkerAssignment {
+    status: String,
+    chunks_deltas: Vec<u64>,
+    encrypted_headers: EncryptedHeaders,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Assignment {
+    datasets: Vec<Dataset>,
+    worker_assignments: HashMap<String, WorkerAssignment>,
+    #[serde(skip)]
+    chunk_map: Option<HashMap<String, u64>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkAssignment {
+    pub(crate) url: String,
+    pub(crate) id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkState {
+    pub(crate) network: String,
+    pub(crate) assignment: NetworkAssignment
+}
+
+impl Assignment {
+    fn add_chunk(&mut self, chunk: Chunk, dataset_id: String, dataset_url: String) {
+        match self.datasets.iter_mut().find(|dataset| dataset.id == dataset_id) {
+            Some(dataset) => dataset.chunks.push(chunk),
+            None => self.datasets.push(Dataset { id: dataset_id, base_url: dataset_url, chunks: vec![chunk] }),
+        }
+        self.chunk_map = None
+    }
+
+    pub fn dataset_chunks_for_peer_id(&self, peer_id: String) -> Option<Vec<Dataset>> {
+        let local_assignment = match self.worker_assignments.get(&peer_id) {
+            Some(worker_assignment) => worker_assignment,
+            None => {
+                return None
+            }
+        };
+        let mut result: Vec<Dataset> = Default::default();
+        let mut idxs: Vec<u64> = Default::default();
+        let mut cursor = 0;
+        for v in &local_assignment.chunks_deltas {
+            cursor += v;
+            idxs.push(cursor);
+        }
+        cursor = 0;
+        for u in &self.datasets {
+            if idxs.is_empty() {
+                break;
+            }
+            let mut filtered_chunks: Vec<Chunk> = Default::default();
+            for v in &u.chunks {
+                if idxs[0] < cursor {
+                    return None; // Malformed diffs
+                }
+                if idxs[0] == cursor {
+                    filtered_chunks.push(v.clone());
+                    idxs.remove(0);
```

Comment:
Why? It's unnecessarily O(n^2). I would do it like this:
```rust
let flat_list: Vec<(&Dataset, &Chunk)> = self.datasets
    .iter()
    .map(|dataset| {
        dataset
            .chunks
            .iter()
            .map(|chunk| (dataset, chunk))
    })
    .flatten()
    .collect();
let mut result: HashMap<String, Dataset> = Default::default();
for index in delta_decode(&local_assignment.chunks_deltas) {
    let (dataset, chunk) = flat_list.get(index as usize)?;
    result.entry(dataset.id).or_insert_with(|| Dataset {
        id: dataset.id.clone(),
        base_url: dataset.base_url.clone(),
        chunks: vec![],
    }).chunks.push((*chunk).clone());
}
Some(result.into_values().collect())
```

BTW, this method should consume `self` and try to reuse the allocated memory

Reply:
fixed